// nrm/dist/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NPMRC = exports.NRMRC = exports.REGISTRY_ATTRS = exports.ALWAYS_AUTH = exports.REPOSITORY = exports.REGISTRY = exports.EMAIL = exports.AUTH = exports.HOME = exports.REGISTRIES = void 0;
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
exports.REGISTRIES = {
    npm: {
        home: 'https://www.npmjs.org',
        registry: 'https://registry.npmjs.org/',
    },
    yarn: {
        home: 'https://yarnpkg.com',
        registry: 'https://registry.yarnpkg.com/',
    },
    tencent: {
        home: 'https://mirrors.tencent.com/npm/',
        registry: 'https://mirrors.tencent.com/npm/',
    },
    cnpm: {
        home: 'https://cnpmjs.org',
        registry: 'https://r.cnpmjs.org/',
    },
    taobao: {
        home: 'https://npmmirror.com',
        registry: 'https://registry.npmmirror.com/',
    },
    npmMirror: {
        home: 'https://skimdb.npmjs.com/',
        registry: 'https://skimdb.npmjs.com/registry/',
    },
    huawei: {
        home: 'https://www.huaweicloud.com/special/npm-jingxiang.html',
        registry: 'https://repo.huaweicloud.com/repository/npm/',
    },
};
exports.HOME = 'home';
exports.AUTH = '_auth';
exports.EMAIL = 'email';
exports.REGISTRY = 'registry';
exports.REPOSITORY = 'repository';
exports.ALWAYS_AUTH = 'always-auth';
exports.REGISTRY_ATTRS = [exports.REGISTRY, exports.HOME, exports.AUTH, exports.ALWAYS_AUTH];
exports.NRMRC = node_path_1.default.join(node_os_1.default.homedir(), '.nrmrc');
exports.NPMRC = node_path_1.default.join(node_os_1.default.homedir(), '.npmrc');
