// @autolabz/llmapi-sdk/dist/client.js
import { createHttpClient } from './http';
import { LLMServiceError } from './errors';
export class LLMAPIClient {
    constructor(config) {
        this.client = createHttpClient(config);
        this.auth = config.auth;
    }
    async health() {
        const res = await this.client.get('/health');
        return res.data;
    }
    async chat(body) {
        var _a, _b;
        // 强制非流式
        const reqBody = { ...body, stream: false };
        const res = await this.client.post('/v1/chat/completions', reqBody);
        return ((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.data) !== null && _b !== void 0 ? _b : res.data);
    }
    async chatStream(body, handlers) {
        var _a, _b, _c, _d, _f, _g, _h, _j, _k, _l, _m;
        // 直接使用 fetch 以便获取可读流；从 axios 实例读取 baseURL 与 headers
        const baseURL = this.client.defaults.baseURL;
        const headers = { ...(((_a = this.client.defaults.headers) === null || _a === void 0 ? void 0 : _a.common) || {}) };
        headers['Accept'] = 'text/event-stream';
        headers['Content-Type'] = 'application/json';
        // axios 默认头中可能包含小写/大写混合，将其展平
        const flatHeaders = {};
        for (const [k, v] of Object.entries(headers)) {
            if (typeof v === 'string')
                flatHeaders[k] = v;
        }
        // 确保附加鉴权相关头（与 axios request 拦截器一致）
        try {
            const [token, clientId] = await Promise.all([
                Promise.resolve((_c = (_b = this.auth) === null || _b === void 0 ? void 0 : _b.getAccessToken) === null || _c === void 0 ? void 0 : _c.call(_b)),
                Promise.resolve((_f = (_d = this.auth) === null || _d === void 0 ? void 0 : _d.getClientId) === null || _f === void 0 ? void 0 : _f.call(_d)),
            ]);
            if (token && !flatHeaders['Authorization']) {
                flatHeaders['Authorization'] = `Bearer ${token}`;
            }
            if (clientId && !flatHeaders['X-Client-Id']) {
                flatHeaders['X-Client-Id'] = String(clientId);
            }
            if (!flatHeaders['x-request-id']) {
                flatHeaders['x-request-id'] = `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
            }
        }
        catch (_e) {
            // 忽略获取 token 失败，后续由服务端返回 401
        }
        // 兼容相对与绝对 baseURL：当 baseURL 为 '/llmapi' 这类相对路径时，URL 构造需要手动拼接
        let url;
        try {
            // 如果 baseURL 是绝对地址（含协议），可直接使用 URL API
            const test = new URL(baseURL);
            url = new URL('/v1/chat/completions', test).toString();
        }
        catch {
            // 相对 base，例如 '/llmapi'，直接按路径拼接，确保只有一个斜杠
            const base = baseURL.endsWith('/') ? baseURL.slice(0, -1) : baseURL;
            url = `${base}/v1/chat/completions`;
        }
        let resp = await fetch(url, {
            method: 'POST',
            headers: flatHeaders,
            body: JSON.stringify({ ...body, stream: true }),
        });
        // 若鉴权失败，尝试刷新一次 token 并重试
        if (resp.status === 401) {
            try {
                const newAccessToken = await Promise.resolve((_h = (_g = this.auth) === null || _g === void 0 ? void 0 : _g.refreshAccessToken) === null || _h === void 0 ? void 0 : _h.call(_g));
                if (newAccessToken) {
                    flatHeaders['Authorization'] = `Bearer ${newAccessToken}`;
                    resp = await fetch(url, {
                        method: 'POST',
                        headers: flatHeaders,
                        body: JSON.stringify({ ...body, stream: true }),
                    });
                }
            }
            catch { }
        }
        if (!resp.ok || !resp.body) {
            const text = await resp.text().catch(() => '');
            const err = new LLMServiceError(`Stream request failed: ${resp.status} ${resp.statusText} ${text}`);
            (_j = handlers === null || handlers === void 0 ? void 0 : handlers.onError) === null || _j === void 0 ? void 0 : _j.call(handlers, err);
            throw err;
        }
        const reader = resp.body.getReader();
        const decoder = new TextDecoder();
        let done = false;
        while (!done) {
            const chunk = await reader.read();
            done = !!chunk.done;
            if (chunk.value) {
                const text = decoder.decode(chunk.value);
                // SSE 以换行分割，逐行解析
                const lines = text.split(/\r?\n/);
                for (const line of lines) {
                    if (!line)
                        continue;
                    (_k = handlers === null || handlers === void 0 ? void 0 : handlers.onEvent) === null || _k === void 0 ? void 0 : _k.call(handlers, line);
                    const m = /^data:\s*(.*)$/.exec(line);
                    if (!m)
                        continue;
                    const payload = m[1];
                    if (payload === '[DONE]') {
                        (_l = handlers === null || handlers === void 0 ? void 0 : handlers.onDone) === null || _l === void 0 ? void 0 : _l.call(handlers);
                        return;
                    }
                    try {
                        const json = JSON.parse(payload);
                        (_m = handlers === null || handlers === void 0 ? void 0 : handlers.onMessage) === null || _m === void 0 ? void 0 : _m.call(handlers, json);
                    }
                    catch (e) {
                        // 忽略不可解析的 data 行（可能是心跳等）
                    }
                }
            }
        }
    }
}
export function createLLMClient(config) {
    return new LLMAPIClient(config);
}
