// @autolabz/data-sdk/dist/errors.js
export class DataServiceError extends Error {
    constructor(message, options) {
        super(message);
        this.name = 'DataServiceError';
        this.httpStatus = options === null || options === void 0 ? void 0 : options.httpStatus;
        this.code = options === null || options === void 0 ? void 0 : options.code;
        this.details = options === null || options === void 0 ? void 0 : options.details;
        this.requestId = options === null || options === void 0 ? void 0 : options.requestId;
    }
    static fromAxiosError(err) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const status = (_a = err.response) === null || _a === void 0 ? void 0 : _a.status;
        const data = (_b = err.response) === null || _b === void 0 ? void 0 : _b.data;
        const requestId = ((_d = (_c = err.response) === null || _c === void 0 ? void 0 : _c.headers) === null || _d === void 0 ? void 0 : _d['x-request-id']) || ((_f = (_e = err.response) === null || _e === void 0 ? void 0 : _e.headers) === null || _f === void 0 ? void 0 : _f['x-requestid']);
        const code = (data === null || data === void 0 ? void 0 : data.code) || ((_g = data === null || data === void 0 ? void 0 : data.error) === null || _g === void 0 ? void 0 : _g.code);
        const message = (data === null || data === void 0 ? void 0 : data.message) || ((_h = data === null || data === void 0 ? void 0 : data.error) === null || _h === void 0 ? void 0 : _h.message) || err.message || 'Request failed';
        return new DataServiceError(message, { httpStatus: status, code, details: data, requestId });
    }
}
