// @autolabz/points-sdk/dist/client.js
import { createHttpClient } from './http';
import { PointsServiceError } from './errors';
export class PointsAPIClient {
    constructor(config) {
        this.client = createHttpClient(config);
    }
    async health() {
        const res = await this.client.get('/health');
        return res.data;
    }
    async getMyBalance() {
        var _a, _b;
        const res = await this.client.get('/v1/points/balance');
        return ((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.data) !== null && _b !== void 0 ? _b : res.data);
    }
    async isBalanceNonNegative() {
        const result = await this.getMyBalance();
        return Number(result.balance) >= 0;
    }
    async consume(body) {
        var _a, _b;
        const amount = Number(body === null || body === void 0 ? void 0 : body.amount);
        if (!Number.isFinite(amount) || !Number.isInteger(amount) || amount <= 0) {
            throw new PointsServiceError('amount must be a positive integer', { code: 'INVALID_ARGUMENT' });
        }
        const res = await this.client.post('/v1/points/consume', body);
        return ((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.data) !== null && _b !== void 0 ? _b : res.data);
    }
    async listMyConsumptions(query) {
        var _a, _b;
        const res = await this.client.get('/v1/points/consumptions', { params: query });
        return ((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.data) !== null && _b !== void 0 ? _b : res.data);
    }
}
export function createPointsClient(config) {
    return new PointsAPIClient(config);
}
