// @autolabz/oauth-sdk/dist/index.mjs
import Je, { useState as ge, useCallback as G, useRef as me, useEffect as re, createContext as Ct, useContext as Pt, useMemo as It } from "react";
import Ve from "axios";
var pe = { exports: {} }, H = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var We;
function At() {
  if (We) return H;
  We = 1;
  var s = Je, r = Symbol.for("react.element"), n = Symbol.for("react.fragment"), i = Object.prototype.hasOwnProperty, u = s.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, f = { key: !0, ref: !0, __self: !0, __source: !0 };
  function l(E, _, y) {
    var g, v = {}, b = null, A = null;
    y !== void 0 && (b = "" + y), _.key !== void 0 && (b = "" + _.key), _.ref !== void 0 && (A = _.ref);
    for (g in _) i.call(_, g) && !f.hasOwnProperty(g) && (v[g] = _[g]);
    if (E && E.defaultProps) for (g in _ = E.defaultProps, _) v[g] === void 0 && (v[g] = _[g]);
    return { $$typeof: r, type: E, key: b, ref: A, props: v, _owner: u.current };
  }
  return H.Fragment = n, H.jsx = l, H.jsxs = l, H;
}
var Q = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ze;
function Ot() {
  return ze || (ze = 1, process.env.NODE_ENV !== "production" && function() {
    var s = Je, r = Symbol.for("react.element"), n = Symbol.for("react.portal"), i = Symbol.for("react.fragment"), u = Symbol.for("react.strict_mode"), f = Symbol.for("react.profiler"), l = Symbol.for("react.provider"), E = Symbol.for("react.context"), _ = Symbol.for("react.forward_ref"), y = Symbol.for("react.suspense"), g = Symbol.for("react.suspense_list"), v = Symbol.for("react.memo"), b = Symbol.for("react.lazy"), A = Symbol.for("react.offscreen"), x = Symbol.iterator, N = "@@iterator";
    function F(e) {
      if (e === null || typeof e != "object")
        return null;
      var t = x && e[x] || e[N];
      return typeof t == "function" ? t : null;
    }
    var j = s.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function S(e) {
      {
        for (var t = arguments.length, a = new Array(t > 1 ? t - 1 : 0), c = 1; c < t; c++)
          a[c - 1] = arguments[c];
        ne("error", e, a);
      }
    }
    function ne(e, t, a) {
      {
        var c = j.ReactDebugCurrentFrame, p = c.getStackAddendum();
        p !== "" && (t += "%s", a = a.concat([p]));
        var k = a.map(function(h) {
          return String(h);
        });
        k.unshift("Warning: " + t), Function.prototype.apply.call(console[e], console, k);
      }
    }
    var ae = !1, o = !1, m = !1, T = !1, C = !1, $;
    $ = Symbol.for("react.module.reference");
    function B(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === i || e === f || C || e === u || e === y || e === g || T || e === A || ae || o || m || typeof e == "object" && e !== null && (e.$$typeof === b || e.$$typeof === v || e.$$typeof === l || e.$$typeof === E || e.$$typeof === _ || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === $ || e.getModuleId !== void 0));
    }
    function Ze(e, t, a) {
      var c = e.displayName;
      if (c)
        return c;
      var p = t.displayName || t.name || "";
      return p !== "" ? a + "(" + p + ")" : a;
    }
    function _e(e) {
      return e.displayName || "Context";
    }
    function Y(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && S("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case i:
          return "Fragment";
        case n:
          return "Portal";
        case f:
          return "Profiler";
        case u:
          return "StrictMode";
        case y:
          return "Suspense";
        case g:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case E:
            var t = e;
            return _e(t) + ".Consumer";
          case l:
            var a = e;
            return _e(a._context) + ".Provider";
          case _:
            return Ze(e, e.render, "ForwardRef");
          case v:
            var c = e.displayName || null;
            return c !== null ? c : Y(e.type) || "Memo";
          case b: {
            var p = e, k = p._payload, h = p._init;
            try {
              return Y(h(k));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var M = Object.assign, J = 0, ye, ke, be, Ee, Re, Te, we;
    function Se() {
    }
    Se.__reactDisabledLog = !0;
    function et() {
      {
        if (J === 0) {
          ye = console.log, ke = console.info, be = console.warn, Ee = console.error, Re = console.group, Te = console.groupCollapsed, we = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: Se,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        J++;
      }
    }
    function tt() {
      {
        if (J--, J === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: M({}, e, {
              value: ye
            }),
            info: M({}, e, {
              value: ke
            }),
            warn: M({}, e, {
              value: be
            }),
            error: M({}, e, {
              value: Ee
            }),
            group: M({}, e, {
              value: Re
            }),
            groupCollapsed: M({}, e, {
              value: Te
            }),
            groupEnd: M({}, e, {
              value: we
            })
          });
        }
        J < 0 && S("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var oe = j.ReactCurrentDispatcher, se;
    function X(e, t, a) {
      {
        if (se === void 0)
          try {
            throw Error();
          } catch (p) {
            var c = p.stack.trim().match(/\n( *(at )?)/);
            se = c && c[1] || "";
          }
        return `
` + se + e;
      }
    }
    var ie = !1, Z;
    {
      var rt = typeof WeakMap == "function" ? WeakMap : Map;
      Z = new rt();
    }
    function Ce(e, t) {
      if (!e || ie)
        return "";
      {
        var a = Z.get(e);
        if (a !== void 0)
          return a;
      }
      var c;
      ie = !0;
      var p = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var k;
      k = oe.current, oe.current = null, et();
      try {
        if (t) {
          var h = function() {
            throw Error();
          };
          if (Object.defineProperty(h.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(h, []);
            } catch (O) {
              c = O;
            }
            Reflect.construct(e, [], h);
          } else {
            try {
              h.call();
            } catch (O) {
              c = O;
            }
            e.call(h.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (O) {
            c = O;
          }
          e();
        }
      } catch (O) {
        if (O && c && typeof O.stack == "string") {
          for (var d = O.stack.split(`
`), I = c.stack.split(`
`), R = d.length - 1, w = I.length - 1; R >= 1 && w >= 0 && d[R] !== I[w]; )
            w--;
          for (; R >= 1 && w >= 0; R--, w--)
            if (d[R] !== I[w]) {
              if (R !== 1 || w !== 1)
                do
                  if (R--, w--, w < 0 || d[R] !== I[w]) {
                    var U = `
` + d[R].replace(" at new ", " at ");
                    return e.displayName && U.includes("<anonymous>") && (U = U.replace("<anonymous>", e.displayName)), typeof e == "function" && Z.set(e, U), U;
                  }
                while (R >= 1 && w >= 0);
              break;
            }
        }
      } finally {
        ie = !1, oe.current = k, tt(), Error.prepareStackTrace = p;
      }
      var K = e ? e.displayName || e.name : "", V = K ? X(K) : "";
      return typeof e == "function" && Z.set(e, V), V;
    }
    function nt(e, t, a) {
      return Ce(e, !1);
    }
    function at(e) {
      var t = e.prototype;
      return !!(t && t.isReactComponent);
    }
    function ee(e, t, a) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return Ce(e, at(e));
      if (typeof e == "string")
        return X(e);
      switch (e) {
        case y:
          return X("Suspense");
        case g:
          return X("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case _:
            return nt(e.render);
          case v:
            return ee(e.type, t, a);
          case b: {
            var c = e, p = c._payload, k = c._init;
            try {
              return ee(k(p), t, a);
            } catch {
            }
          }
        }
      return "";
    }
    var q = Object.prototype.hasOwnProperty, Pe = {}, Ie = j.ReactDebugCurrentFrame;
    function te(e) {
      if (e) {
        var t = e._owner, a = ee(e.type, e._source, t ? t.type : null);
        Ie.setExtraStackFrame(a);
      } else
        Ie.setExtraStackFrame(null);
    }
    function ot(e, t, a, c, p) {
      {
        var k = Function.call.bind(q);
        for (var h in e)
          if (k(e, h)) {
            var d = void 0;
            try {
              if (typeof e[h] != "function") {
                var I = Error((c || "React class") + ": " + a + " type `" + h + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[h] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw I.name = "Invariant Violation", I;
              }
              d = e[h](t, h, c, a, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (R) {
              d = R;
            }
            d && !(d instanceof Error) && (te(p), S("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", c || "React class", a, h, typeof d), te(null)), d instanceof Error && !(d.message in Pe) && (Pe[d.message] = !0, te(p), S("Failed %s type: %s", a, d.message), te(null));
          }
      }
    }
    var st = Array.isArray;
    function ce(e) {
      return st(e);
    }
    function it(e) {
      {
        var t = typeof Symbol == "function" && Symbol.toStringTag, a = t && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return a;
      }
    }
    function ct(e) {
      try {
        return Ae(e), !1;
      } catch {
        return !0;
      }
    }
    function Ae(e) {
      return "" + e;
    }
    function Oe(e) {
      if (ct(e))
        return S("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", it(e)), Ae(e);
    }
    var je = j.ReactCurrentOwner, lt = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, xe, Fe;
    function ut(e) {
      if (q.call(e, "ref")) {
        var t = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (t && t.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function ft(e) {
      if (q.call(e, "key")) {
        var t = Object.getOwnPropertyDescriptor(e, "key").get;
        if (t && t.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function dt(e, t) {
      typeof e.ref == "string" && je.current;
    }
    function ht(e, t) {
      {
        var a = function() {
          xe || (xe = !0, S("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", t));
        };
        a.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: a,
          configurable: !0
        });
      }
    }
    function vt(e, t) {
      {
        var a = function() {
          Fe || (Fe = !0, S("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", t));
        };
        a.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: a,
          configurable: !0
        });
      }
    }
    var gt = function(e, t, a, c, p, k, h) {
      var d = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: r,
        // Built-in properties that belong on the element
        type: e,
        key: t,
        ref: a,
        props: h,
        // Record the component responsible for creating this element.
        _owner: k
      };
      return d._store = {}, Object.defineProperty(d._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(d, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: c
      }), Object.defineProperty(d, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: p
      }), Object.freeze && (Object.freeze(d.props), Object.freeze(d)), d;
    };
    function mt(e, t, a, c, p) {
      {
        var k, h = {}, d = null, I = null;
        a !== void 0 && (Oe(a), d = "" + a), ft(t) && (Oe(t.key), d = "" + t.key), ut(t) && (I = t.ref, dt(t, p));
        for (k in t)
          q.call(t, k) && !lt.hasOwnProperty(k) && (h[k] = t[k]);
        if (e && e.defaultProps) {
          var R = e.defaultProps;
          for (k in R)
            h[k] === void 0 && (h[k] = R[k]);
        }
        if (d || I) {
          var w = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          d && ht(h, w), I && vt(h, w);
        }
        return gt(e, d, I, p, c, je.current, h);
      }
    }
    var le = j.ReactCurrentOwner, $e = j.ReactDebugCurrentFrame;
    function z(e) {
      if (e) {
        var t = e._owner, a = ee(e.type, e._source, t ? t.type : null);
        $e.setExtraStackFrame(a);
      } else
        $e.setExtraStackFrame(null);
    }
    var ue;
    ue = !1;
    function fe(e) {
      return typeof e == "object" && e !== null && e.$$typeof === r;
    }
    function Ue() {
      {
        if (le.current) {
          var e = Y(le.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function pt(e) {
      return "";
    }
    var Le = {};
    function _t(e) {
      {
        var t = Ue();
        if (!t) {
          var a = typeof e == "string" ? e : e.displayName || e.name;
          a && (t = `

Check the top-level render call using <` + a + ">.");
        }
        return t;
      }
    }
    function De(e, t) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var a = _t(t);
        if (Le[a])
          return;
        Le[a] = !0;
        var c = "";
        e && e._owner && e._owner !== le.current && (c = " It was passed a child from " + Y(e._owner.type) + "."), z(e), S('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', a, c), z(null);
      }
    }
    function Ne(e, t) {
      {
        if (typeof e != "object")
          return;
        if (ce(e))
          for (var a = 0; a < e.length; a++) {
            var c = e[a];
            fe(c) && De(c, t);
          }
        else if (fe(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var p = F(e);
          if (typeof p == "function" && p !== e.entries)
            for (var k = p.call(e), h; !(h = k.next()).done; )
              fe(h.value) && De(h.value, t);
        }
      }
    }
    function yt(e) {
      {
        var t = e.type;
        if (t == null || typeof t == "string")
          return;
        var a;
        if (typeof t == "function")
          a = t.propTypes;
        else if (typeof t == "object" && (t.$$typeof === _ || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        t.$$typeof === v))
          a = t.propTypes;
        else
          return;
        if (a) {
          var c = Y(t);
          ot(a, e.props, "prop", c, e);
        } else if (t.PropTypes !== void 0 && !ue) {
          ue = !0;
          var p = Y(t);
          S("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", p || "Unknown");
        }
        typeof t.getDefaultProps == "function" && !t.getDefaultProps.isReactClassApproved && S("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function kt(e) {
      {
        for (var t = Object.keys(e.props), a = 0; a < t.length; a++) {
          var c = t[a];
          if (c !== "children" && c !== "key") {
            z(e), S("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", c), z(null);
            break;
          }
        }
        e.ref !== null && (z(e), S("Invalid attribute `ref` supplied to `React.Fragment`."), z(null));
      }
    }
    var Ye = {};
    function Me(e, t, a, c, p, k) {
      {
        var h = B(e);
        if (!h) {
          var d = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (d += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var I = pt();
          I ? d += I : d += Ue();
          var R;
          e === null ? R = "null" : ce(e) ? R = "array" : e !== void 0 && e.$$typeof === r ? (R = "<" + (Y(e.type) || "Unknown") + " />", d = " Did you accidentally export a JSX literal instead of a component?") : R = typeof e, S("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", R, d);
        }
        var w = mt(e, t, a, p, k);
        if (w == null)
          return w;
        if (h) {
          var U = t.children;
          if (U !== void 0)
            if (c)
              if (ce(U)) {
                for (var K = 0; K < U.length; K++)
                  Ne(U[K], e);
                Object.freeze && Object.freeze(U);
              } else
                S("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Ne(U, e);
        }
        if (q.call(t, "key")) {
          var V = Y(e), O = Object.keys(t).filter(function(St) {
            return St !== "key";
          }), de = O.length > 0 ? "{key: someKey, " + O.join(": ..., ") + ": ...}" : "{key: someKey}";
          if (!Ye[V + de]) {
            var wt = O.length > 0 ? "{" + O.join(": ..., ") + ": ...}" : "{}";
            S(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`, de, V, wt, V), Ye[V + de] = !0;
          }
        }
        return e === i ? kt(w) : yt(w), w;
      }
    }
    function bt(e, t, a) {
      return Me(e, t, a, !0);
    }
    function Et(e, t, a) {
      return Me(e, t, a, !1);
    }
    var Rt = Et, Tt = bt;
    Q.Fragment = i, Q.jsx = Rt, Q.jsxs = Tt;
  }()), Q;
}
process.env.NODE_ENV === "production" ? pe.exports = At() : pe.exports = Ot();
var P = pe.exports;
const he = "autolab_oauth_access_token", ve = "autolab_oauth_refresh_token";
class jt {
  constructor(r) {
    this.isRefreshing = !1, this.failedQueue = [], this.onTokenRefreshed = r.onTokenRefreshed, this.onRefreshFailed = r.onRefreshFailed, this.clientIdProvider = r.clientIdProvider, this.client = Ve.create({
      baseURL: r.baseURL,
      headers: { "Content-Type": "application/json" }
    }), this.setupInterceptors();
  }
  setupInterceptors() {
    this.client.interceptors.request.use(
      (r) => {
        const n = localStorage.getItem(he);
        return n && !r.headers.Authorization && (r.headers.Authorization = `Bearer ${n}`), r;
      },
      (r) => Promise.reject(r)
    ), this.client.interceptors.response.use(
      (r) => r,
      async (r) => {
        var i, u, f;
        const n = r.config;
        if (((i = r.response) == null ? void 0 : i.status) === 401 && !n._retry) {
          if ((u = n.url) != null && u.includes("/oauth/token") || (f = n.url) != null && f.includes("/oauth/revoke"))
            return Promise.reject(r);
          if (this.isRefreshing)
            return new Promise((l, E) => {
              this.failedQueue.push({ resolve: l, reject: E });
            }).then((l) => (n.headers && (n.headers.Authorization = `Bearer ${l}`), this.client(n))).catch((l) => Promise.reject(l));
          n._retry = !0, this.isRefreshing = !0;
          try {
            const l = await this.performTokenRefresh();
            return n.headers && (n.headers.Authorization = `Bearer ${l.accessToken}`), this.processQueue(null, l.accessToken), this.client(n);
          } catch (l) {
            return this.processQueue(l, null), this.onRefreshFailed && this.onRefreshFailed(), Promise.reject(l);
          } finally {
            this.isRefreshing = !1;
          }
        }
        return Promise.reject(r);
      }
    );
  }
  processQueue(r, n = null) {
    this.failedQueue.forEach((i) => {
      r ? i.reject(r) : i.resolve(n);
    }), this.failedQueue = [];
  }
  async performTokenRefresh() {
    const r = localStorage.getItem(ve);
    if (!r) throw new Error("No refresh token available");
    const n = await this.refreshToken(r);
    return localStorage.setItem(he, n.accessToken), localStorage.setItem(ve, n.refreshToken), this.onTokenRefreshed && this.onTokenRefreshed(n.accessToken, n.refreshToken), n;
  }
  async revoke(r) {
    var n;
    try {
      await this.client.post("/oauth/revoke", {
        token: r,
        token_type_hint: "access_token",
        ...(n = this.clientIdProvider) != null && n.call(this) ? { client_id: this.clientIdProvider() } : {}
      });
    } catch {
    }
  }
  async revokeAll() {
    var i, u;
    const r = localStorage.getItem(ve), n = localStorage.getItem(he);
    try {
      n && await this.client.post("/oauth/revoke", {
        token: n,
        token_type_hint: "access_token",
        ...(i = this.clientIdProvider) != null && i.call(this) ? { client_id: this.clientIdProvider() } : {}
      }), r && await this.client.post("/oauth/revoke", {
        token: r,
        token_type_hint: "refresh_token",
        ...(u = this.clientIdProvider) != null && u.call(this) ? { client_id: this.clientIdProvider() } : {}
      });
    } catch {
    }
  }
  async refreshToken(r) {
    var i, u, f;
    const n = await Ve.post(`${this.client.defaults.baseURL}/oauth/token`, {
      grant_type: "refresh_token",
      refresh_token: r,
      ...(i = this.clientIdProvider) != null && i.call(this) ? { client_id: this.clientIdProvider() } : {}
    }, {
      headers: { "Content-Type": "application/json" }
    });
    if (!((u = n.data) != null && u.access_token) || !((f = n.data) != null && f.refresh_token))
      throw new Error("Invalid refresh token response");
    return {
      accessToken: n.data.access_token,
      refreshToken: n.data.refresh_token
    };
  }
  getClient() {
    return this.client;
  }
  async refreshAccessToken() {
    return (await this.performTokenRefresh()).accessToken;
  }
}
const qe = "autolab_pkce_verifier_", Ke = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
function Ge() {
  try {
    if (typeof globalThis < "u" && globalThis.crypto) return globalThis.crypto;
  } catch {
  }
}
function He(s) {
  const r = Ge(), n = new Uint8Array(s);
  if (r && r.getRandomValues)
    r.getRandomValues(n);
  else
    for (let i = 0; i < s; i++)
      n[i] = Math.floor(Math.random() * 256);
  return n;
}
function xt(s = 64) {
  const r = Math.min(128, Math.max(43, s)), n = He(r);
  let i = "";
  for (let u = 0; u < n.length; u++)
    i += Ke[n[u] % Ke.length];
  return i;
}
function Ft(s) {
  const r = new Uint8Array(s);
  let n = "";
  for (let u = 0; u < r.byteLength; u++)
    n += String.fromCharCode(r[u]);
  return btoa(n).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}
async function $t(s) {
  const r = Ge();
  if (!r || !r.subtle)
    throw new Error("Web Crypto not available: S256 is required");
  const n = new TextEncoder().encode(s), i = await r.subtle.digest("SHA-256", n);
  return Ft(i);
}
function Ut(s, r) {
  try {
    sessionStorage.setItem(qe + s, r);
  } catch {
  }
}
function Lt(s) {
  const r = qe + s;
  try {
    const n = sessionStorage.getItem(r);
    return sessionStorage.removeItem(r), n;
  } catch {
    return null;
  }
}
function Dt(s = 24) {
  const r = He(s);
  let n = "";
  for (let u = 0; u < r.byteLength; u++) n += String.fromCharCode(r[u]);
  return btoa(n).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}
function Nt(s, r) {
  const n = s.replace(/\/$/, ""), i = r.startsWith("/") ? r : `/${r}`;
  return `${n}${i}`;
}
async function Yt(s) {
  const { authServiceUrl: r, clientId: n, redirectUri: i, scope: u, usePkce: f = !0, additionalParams: l } = s, E = s.state || Dt(24);
  let _;
  if (f) {
    const g = xt();
    Ut(E, g), _ = await $t(g);
  }
  const y = new URL(Nt(r, "/oauth/authorize"));
  if (y.searchParams.set("response_type", "code"), y.searchParams.set("client_id", n), y.searchParams.set("redirect_uri", i), y.searchParams.set("state", E), u && y.searchParams.set("scope", u), f && _ && (y.searchParams.set("code_challenge", _), y.searchParams.set("code_challenge_method", "S256")), l) {
    const g = /* @__PURE__ */ new Set(["response_type", "client_id", "redirect_uri", "state", "scope", "code_challenge", "code_challenge_method"]);
    for (const [v, b] of Object.entries(l))
      g.has(v) || y.searchParams.set(v, String(b));
  }
  return y;
}
const Mt = "autolab_oauth_access_token", Vt = "autolab_oauth_refresh_token";
async function Wt(s) {
  return Yt(s);
}
async function zt(s) {
  const r = await Wt(s);
  window.location.assign(r.toString());
}
function Be(s, r) {
  const n = s.replace(/\/$/, ""), i = r.startsWith("/") ? r : `/${r}`;
  return `${n}${i}`;
}
async function Kt(s) {
  const { authServiceUrl: r, clientId: n, redirectUri: i, fetchUserinfo: u = !0 } = s, f = new URLSearchParams(window.location.search), l = f.get("error"), E = f.get("error_description");
  if (l)
    throw new Error(`OAuth error: ${l}${E ? ` - ${E}` : ""}`);
  const _ = f.get("code"), y = f.get("state") || "";
  if (!_ || !y)
    throw new Error("Missing authorization code or state in redirect URL");
  const g = Lt(y);
  if (!g)
    throw new Error("Missing PKCE verifier for the given state");
  const v = await fetch(Be(r, "/oauth/token"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      grant_type: "authorization_code",
      code: _,
      redirect_uri: i,
      client_id: n,
      code_verifier: g
    })
  });
  if (!v.ok) {
    const F = await v.text().catch(() => "");
    throw new Error(`Token exchange failed: ${v.status} ${F}`);
  }
  const b = await v.json(), A = b.access_token, x = b.refresh_token;
  if (!A)
    throw new Error("Invalid token response");
  try {
    localStorage.setItem(Mt, A), x && localStorage.setItem(Vt, x);
  } catch {
  }
  let N;
  if (u)
    try {
      const F = await fetch(Be(r, "/oauth/userinfo"), {
        headers: { Authorization: `Bearer ${A}` }
      });
      F.ok && (N = await F.json());
    } catch {
    }
  return { accessToken: A, refreshToken: x, userinfo: N, state: y };
}
const Qe = Ct(void 0), D = "autolab_oauth_state", W = "autolab_oauth_refresh_token", L = "autolab_oauth_access_token";
function qt({ children: s, authServiceUrl: r, clientId: n, clientIdProvider: i }) {
  const [u, f] = ge({
    isAuthenticated: !1,
    user: null,
    accessToken: null
  }), [l, E] = ge(!1), _ = G((o, m) => {
    f((T) => ({
      ...T,
      accessToken: o
    }));
  }, []), y = G(() => {
    localStorage.removeItem(D), localStorage.removeItem(L), localStorage.removeItem(W), f({ isAuthenticated: !1, user: null, accessToken: null });
  }, []), g = me(null), v = G(() => {
    if (n) return n;
    if (i)
      try {
        const o = i();
        if (o) return o;
      } catch {
      }
    try {
      const o = new URLSearchParams(window.location.search), m = o.get("client_id") || o.get("clientId");
      if (m)
        return console.warn("[autolab][oauth-sdk] clientId resolved from URL query. For production, pass a fixed clientId via <OAuthProvider clientId=...> or clientIdProvider()."), m;
    } catch {
    }
    try {
      const o = sessionStorage.getItem("autolab_client_id") || null;
      if (o)
        return console.warn("[autolab][oauth-sdk] clientId resolved from sessionStorage. For production, pass a fixed clientId via <OAuthProvider clientId=...> or clientIdProvider()."), o;
    } catch {
    }
    return null;
  }, [n, i]);
  g.current || (g.current = new jt({
    baseURL: r,
    onTokenRefreshed: _,
    onRefreshFailed: y,
    clientIdProvider: v
  }));
  const b = g.current, A = me(!1);
  re(() => {
    const o = localStorage.getItem(D), m = localStorage.getItem(L);
    if (o)
      try {
        const T = JSON.parse(o);
        f({
          ...T,
          accessToken: m
        });
      } catch (T) {
        console.error("Failed to parse auth state:", T);
      }
    E(!0);
  }, []), re(() => {
    if (!l || A.current) return;
    A.current = !0;
    let o = !1;
    async function m() {
      const T = localStorage.getItem(L), C = localStorage.getItem(W);
      if (!(!T && !C)) {
        if (T)
          try {
            await b.getClient().get("/oauth/userinfo");
            return;
          } catch {
          }
        if (C)
          try {
            const $ = await b.refreshAccessToken();
            o || f((B) => ({ ...B, isAuthenticated: !0, accessToken: $ }));
            return;
          } catch {
          }
        o || (localStorage.removeItem(D), localStorage.removeItem(L), localStorage.removeItem(W), f({ isAuthenticated: !1, user: null, accessToken: null }));
      }
    }
    return m(), () => {
      o = !0;
    };
  }, [l, b]), re(() => {
    const o = (m) => {
      if (m.key === D && m.newValue)
        try {
          const T = JSON.parse(m.newValue), C = localStorage.getItem(L);
          f({
            ...T,
            accessToken: C
          });
        } catch (T) {
          console.error("Failed to parse storage event:", T);
        }
      else m.key === D && !m.newValue && f({ isAuthenticated: !1, user: null, accessToken: null });
    };
    return window.addEventListener("storage", o), () => window.removeEventListener("storage", o);
  }, []);
  const x = (o, m, T) => {
    const C = {
      isAuthenticated: !0,
      user: o,
      accessToken: m
    }, $ = { isAuthenticated: !0, user: o };
    localStorage.setItem(D, JSON.stringify($)), localStorage.setItem(L, m), localStorage.setItem(W, T), f(C);
  }, N = async () => {
    if (u.accessToken || localStorage.getItem(L))
      try {
        await b.revokeAll();
      } catch (m) {
        console.error("Revoke failed:", m);
      }
    localStorage.removeItem(D), localStorage.removeItem(L), localStorage.removeItem(W), f({ isAuthenticated: !1, user: null, accessToken: null });
  }, F = () => {
    const o = localStorage.getItem(D), m = localStorage.getItem(L);
    if (o) {
      const T = JSON.parse(o);
      f({
        ...T,
        accessToken: m
      });
    }
  }, j = () => u.accessToken || localStorage.getItem(L), S = (o) => {
    localStorage.setItem(L, o), f((m) => ({
      ...m,
      accessToken: o
    }));
  }, ne = G(async (o) => {
    const m = o.clientIdOverride || v();
    if (!m) throw new Error("Missing client_id");
    await zt({
      authServiceUrl: r,
      clientId: m,
      redirectUri: o.redirectUri,
      scope: o.scope,
      state: o.state,
      usePkce: o.usePkce ?? !0,
      additionalParams: o.additionalParams
    });
  }, [r, v]), ae = G(async (o) => {
    const m = (o == null ? void 0 : o.clientIdOverride) || v();
    if (!m) throw new Error("Missing client_id");
    const T = (o == null ? void 0 : o.redirectUri) || `${window.location.origin}${window.location.pathname}`, C = await Kt({
      authServiceUrl: r,
      clientId: m,
      redirectUri: T,
      fetchUserinfo: (o == null ? void 0 : o.fetchUserinfo) ?? !0
    });
    if (C.userinfo) {
      const $ = C.userinfo, B = C.refreshToken || localStorage.getItem(W) || "";
      x($, C.accessToken, B);
    } else {
      localStorage.setItem(L, C.accessToken), C.refreshToken && localStorage.setItem(W, C.refreshToken);
      const $ = { isAuthenticated: !0, user: null };
      localStorage.setItem(D, JSON.stringify($)), f({ isAuthenticated: !0, user: null, accessToken: C.accessToken });
    }
    return C;
  }, [r, v, x]);
  return /* @__PURE__ */ P.jsx(
    Qe.Provider,
    {
      value: {
        ...u,
        isInitialized: l,
        login: x,
        logout: N,
        refreshAuth: F,
        getAccessToken: j,
        updateAccessToken: S,
        getClientId: v,
        apiClient: b,
        startLogin: ne,
        handleRedirect: ae
      },
      children: s
    }
  );
}
function Xe() {
  const s = Pt(Qe);
  if (!s)
    throw new Error("useOAuth must be used within OAuthProvider");
  return s;
}
function Gt({ redirectUri: s, scope: r, usePkce: n, additionalParams: i, profileUrl: u }) {
  const { isAuthenticated: f, user: l, logout: E, startLogin: _ } = Xe(), [y, g] = ge(!1), v = me(null);
  re(() => {
    const j = (S) => {
      v.current && !v.current.contains(S.target) && g(!1);
    };
    return document.addEventListener("mousedown", j), () => document.removeEventListener("mousedown", j);
  }, []);
  const b = async () => {
    try {
      await _({
        redirectUri: s,
        scope: r || "openid profile email",
        usePkce: typeof n == "boolean" ? n : !0,
        additionalParams: i
      });
    } catch (j) {
      console.error("AuthAvatar OAuth login failed:", j);
    }
  }, A = u || `${window.location.origin}/auth/profile`, x = () => {
    window.location.href = A;
  }, N = async () => {
    await E(), g(!1);
  };
  if (!f)
    return /* @__PURE__ */ P.jsx(
      "button",
      {
        className: "autolab-auth-login-btn",
        onClick: b,
        title: "登录",
        "aria-label": "登录",
        children: /* @__PURE__ */ P.jsxs(
          "svg",
          {
            width: "16",
            height: "16",
            viewBox: "0 0 24 24",
            fill: "none",
            stroke: "currentColor",
            strokeWidth: "2",
            strokeLinecap: "round",
            strokeLinejoin: "round",
            children: [
              /* @__PURE__ */ P.jsx("path", { d: "M15 3h4a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2h-4" }),
              /* @__PURE__ */ P.jsx("polyline", { points: "10 17 15 12 10 7" }),
              /* @__PURE__ */ P.jsx("line", { x1: "15", y1: "12", x2: "3", y2: "12" })
            ]
          }
        )
      }
    );
  const F = (l == null ? void 0 : l.avatarUrl) || `https://ui-avatars.com/api/?name=${(l == null ? void 0 : l.nickname) || "User"}&background=1890ff&color=fff`;
  return /* @__PURE__ */ P.jsxs("div", { className: "autolab-auth-avatar", ref: v, children: [
    /* @__PURE__ */ P.jsx(
      "img",
      {
        src: F,
        alt: "User Avatar",
        className: "autolab-auth-avatar-img",
        onClick: () => g(!y)
      }
    ),
    y && /* @__PURE__ */ P.jsxs("div", { className: "autolab-auth-menu", children: [
      /* @__PURE__ */ P.jsxs("div", { className: "autolab-auth-menu-item autolab-auth-user-info", children: [
        /* @__PURE__ */ P.jsx("div", { className: "autolab-auth-nickname", children: (l == null ? void 0 : l.nickname) || "未设置昵称" }),
        /* @__PURE__ */ P.jsx("div", { className: "autolab-auth-email", children: l == null ? void 0 : l.email })
      ] }),
      /* @__PURE__ */ P.jsx("div", { className: "autolab-auth-menu-divider" }),
      /* @__PURE__ */ P.jsx("button", { className: "autolab-auth-menu-item", onClick: x, children: "个人中心" }),
      /* @__PURE__ */ P.jsx("button", { className: "autolab-auth-menu-item", onClick: N, children: "退出登录" })
    ] })
  ] });
}
function Ht({
  redirectUri: s,
  scope: r,
  usePkce: n = !0,
  additionalParams: i,
  label: u,
  className: f,
  onError: l
}) {
  const { startLogin: E } = Xe(), _ = It(() => u || "登录", [u]), y = async (g) => {
    g.preventDefault();
    try {
      await E({
        redirectUri: s,
        scope: r || "openid profile email",
        usePkce: n,
        additionalParams: i
      });
    } catch (v) {
      const b = v instanceof Error ? v : new Error(String(v));
      l ? l(b) : console.error("OAuth login failed:", b);
    }
  };
  return /* @__PURE__ */ P.jsx("button", { className: f, onClick: y, children: _ });
}
function Qt(s) {
  return {
    getAccessToken: () => s.getAccessToken(),
    getClientId: () => s.getClientId(),
    refreshAccessToken: () => s.apiClient.refreshAccessToken()
  };
}
function Xt(s, r) {
  return {
    getAccessToken: r.getAccessToken,
    getClientId: r.getClientId,
    refreshAccessToken: () => s.refreshAccessToken(),
    onUnauthorized: r.onUnauthorized
  };
}
export {
  Gt as AuthAvatar,
  jt as OAuthAPIClient,
  Ht as OAuthLoginButton,
  qt as OAuthProvider,
  Yt as buildAuthorizeUrl,
  Xt as createAuthBridgeFromClient,
  Qt as createAuthBridgeFromContext,
  Wt as getAuthorizeUrl,
  Kt as handleRedirectCallback,
  zt as startAuthorization,
  Xe as useOAuth
};
